File: libs/moqtail-ts/src/model/control/fetch.ts
import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { Location } from '../common/location'
import { KeyValuePair } from '../common/pair'
import { ControlMessageType, FetchType, GroupOrder } from './constant'
import { LengthExceedsMaxError, NotEnoughBytesError } from '../error/error'
import { FullTrackName } from '../data'

export class Fetch {
  constructor(
    public readonly requestId: bigint,
    public readonly subscriberPriority: number,
    public readonly groupOrder: GroupOrder,
    public readonly typeAndProps:
      | {
          readonly type: FetchType.StandAlone
          readonly props: { fullTrackName: FullTrackName; startLocation: Location; endLocation: Location }
        }
      | {
          readonly type: FetchType.Relative
          readonly props: { joiningRequestId: bigint; joiningStart: bigint }
        }
      | {
          readonly type: FetchType.Absolute
          readonly props: { joiningRequestId: bigint; joiningStart: bigint }
        },
    public readonly parameters: KeyValuePair[],
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.Fetch
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.Fetch)
    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putU8(this.subscriberPriority)
    payload.putU8(this.groupOrder)
    payload.putVI(this.typeAndProps.type)
    switch (this.typeAndProps.type) {
      case FetchType.Absolute:
      case FetchType.Relative: {
        payload.putVI(this.typeAndProps.props.joiningRequestId)
        payload.putVI(this.typeAndProps.props.joiningStart)
        break
      }
      case FetchType.StandAlone: {
        payload.putFullTrackName(this.typeAndProps.props.fullTrackName)
        payload.putLocation(this.typeAndProps.props.startLocation)
        payload.putLocation(this.typeAndProps.props.endLocation)
        break
      }
    }
    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putKeyValuePair(param)
    }
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('Fetch::serialize(payload_length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): Fetch {
    const requestId = buf.getVI()
    if (buf.remaining < 1) throw new NotEnoughBytesError('Fetch::parse_payload(subscriber_priority)', 1, buf.remaining)
    const subscriberPriority = buf.getU8()
    if (buf.remaining < 1) throw new NotEnoughBytesError('Fetch::parse_payload(group_order)', 1, buf.remaining)
    const groupOrderRaw = buf.getU8()
    const groupOrder = GroupOrder.tryFrom(groupOrderRaw)
    const fetchTypeRaw = buf.getVI()
    const fetchType = FetchType.tryFrom(fetchTypeRaw)

    let props: Fetch['typeAndProps']

    switch (fetchType) {
      case FetchType.Absolute:
      case FetchType.Relative: {
        const joiningRequestId = buf.getVI()
        const joiningStart = buf.getVI()
        props = {
          type: fetchType,
          props: { joiningRequestId, joiningStart },
        }
        break
      }
      case FetchType.StandAlone: {
        const fullTrackName = buf.getFullTrackName()
        const startLocation = buf.getLocation()
        const endLocation = buf.getLocation()
        props = {
          type: FetchType.StandAlone,
          props: { fullTrackName, startLocation, endLocation },
        }
        break
      }
      default:
        throw new Error(`Unknown fetch type: ${fetchType}`)
    }

    const paramCount = buf.getNumberVI()
    const parameters: KeyValuePair[] = new Array(paramCount)
    for (let i = 0; i < paramCount; i++) {
      parameters[i] = buf.getKeyValuePair()
    }

    return new Fetch(requestId, subscriberPriority, groupOrder, props, parameters)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('Fetch', () => {
    test('roundtrip', () => {
      const requestId = 161803n
      const subscriberPriority = 15
      const groupOrder = GroupOrder.Descending
      const parameters = [
        KeyValuePair.tryNewVarInt(4444, 12321n),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('fetch me ok')),
      ]
      const fetch = new Fetch(
        requestId,
        subscriberPriority,
        groupOrder,
        {
          type: FetchType.Absolute,
          props: { joiningRequestId: 119n, joiningStart: 73n },
        },
        parameters,
      )
      const serialized = fetch.serialize()
      const buf = new ByteBuffer()
      buf.putBytes(serialized.toUint8Array())
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Fetch))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = Fetch.parsePayload(frozen)
      expect(deserialized).toEqual(fetch)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const requestId = 161803n
      const subscriberPriority = 15
      const groupOrder = GroupOrder.Descending
      const parameters = [
        KeyValuePair.tryNewVarInt(4444, 12321n),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('fetch me ok')),
      ]
      const fetch = new Fetch(
        requestId,
        subscriberPriority,
        groupOrder,
        {
          type: FetchType.Absolute,
          props: { joiningRequestId: 119n, joiningStart: 73n },
        },
        parameters,
      )
      const serialized = fetch.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.Fetch))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = Fetch.parsePayload(frozen)
      expect(deserialized).toEqual(fetch)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const requestId = 161803n
      const subscriberPriority = 15
      const groupOrder = GroupOrder.Descending
      const parameters = [
        KeyValuePair.tryNewVarInt(4444, 12321n),
        KeyValuePair.tryNewBytes(1, new TextEncoder().encode('fetch me ok')),
      ]
      const fetch = new Fetch(
        requestId,
        subscriberPriority,
        groupOrder,
        {
          type: FetchType.Absolute,
          props: { joiningRequestId: 119n, joiningStart: 73n },
        },
        parameters,
      )
      const serialized = fetch.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        Fetch.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/fetch.rs
use super::constant::{ControlMessageType, FetchType, GroupOrder};
use super::control_message::ControlMessageTrait;
use crate::model::common::location::Location;
use crate::model::common::pair::KeyValuePair;
use crate::model::common::tuple::Tuple;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use bytes::{Buf, BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct StandAloneFetchProps {
  pub track_namespace: Tuple,
  pub track_name: String,
  pub start_location: Location,
  pub end_location: Location,
}

#[derive(Debug, PartialEq, Clone)]
pub struct JoiningFetchProps {
  pub joining_request_id: u64,
  pub joining_start: u64,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Fetch {
  pub request_id: u64,
  pub subscriber_priority: u8,
  pub group_order: GroupOrder,
  pub fetch_type: FetchType,
  pub standalone_fetch_props: Option<StandAloneFetchProps>,
  pub joining_fetch_props: Option<JoiningFetchProps>,
  pub parameters: Vec<KeyValuePair>,
}

impl Fetch {
  pub fn new(
    request_id: u64,
    subscriber_priority: u8,
    group_order: GroupOrder,
    fetch_type: FetchType,
    standalone_fetch_props: Option<StandAloneFetchProps>,
    joining_fetch_props: Option<JoiningFetchProps>,
    parameters: Vec<KeyValuePair>,
  ) -> Self {
    Self {
      request_id,
      subscriber_priority,
      group_order,
      fetch_type,
      standalone_fetch_props,
      joining_fetch_props,
      parameters,
    }
  }
}

impl ControlMessageTrait for Fetch {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::Fetch as u64)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;
    payload.put_u8(self.subscriber_priority);
    payload.put_u8(self.group_order as u8);
    payload.put_vi(self.fetch_type)?;
    match &self.fetch_type {
      FetchType::AbsoluteFetch => {
        let props = self.joining_fetch_props.as_ref().unwrap();
        payload.put_vi(props.joining_request_id)?;
        payload.put_vi(props.joining_start)?;
      }
      FetchType::RelativeFetch => {
        let props = self.joining_fetch_props.as_ref().unwrap();
        payload.put_vi(props.joining_request_id)?;
        payload.put_vi(props.joining_start)?;
      }
      FetchType::StandAlone => {
        let props = self.standalone_fetch_props.as_ref().unwrap();
        payload.extend_from_slice(&props.track_namespace.serialize()?);
        payload.put_vi(props.track_name.len())?;
        payload.extend_from_slice(props.track_name.as_bytes());
        payload.extend_from_slice(&props.start_location.serialize()?);
        payload.extend_from_slice(&props.end_location.serialize()?);
      }
    }

    payload.put_vi(self.parameters.len())?;
    for param in &self.parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "Fetch::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;

    if payload.remaining() < 1 {
      return Err(ParseError::NotEnoughBytes {
        context: "Fetch::parse_payload(subscriber_priority)",
        needed: 1,
        available: 0,
      });
    }
    let subscriber_priority = payload.get_u8();

    if payload.remaining() < 1 {
      return Err(ParseError::NotEnoughBytes {
        context: "Fetch::parse_payload(group_order)",
        needed: 1,
        available: 0,
      });
    }
    let group_order_raw = payload.get_u8();
    let group_order = GroupOrder::try_from(group_order_raw)?;

    let fetch_type_raw = payload.get_vi()?;
    let fetch_type = FetchType::try_from(fetch_type_raw)?;

    let mut standalone_fetch_props: Option<StandAloneFetchProps> = None;
    let mut joining_fetch_props: Option<JoiningFetchProps> = None;

    match fetch_type {
      FetchType::AbsoluteFetch => {
        let joining_request_id = payload.get_vi()?;
        let joining_start = payload.get_vi()?;
        joining_fetch_props = Some(JoiningFetchProps {
          joining_request_id,
          joining_start,
        });
      }
      FetchType::RelativeFetch => {
        let joining_request_id = payload.get_vi()?;
        let joining_start = payload.get_vi()?;
        joining_fetch_props = Some(JoiningFetchProps {
          joining_request_id,
          joining_start,
        });
      }
      FetchType::StandAlone => {
        let track_namespace = Tuple::deserialize(payload)?;
        let track_name_len = payload.get_vi()?;
        let mut track_name_bytes = vec![0u8; track_name_len as usize];
        payload.copy_to_slice(&mut track_name_bytes);
        let track_name =
          String::from_utf8(track_name_bytes).map_err(|e| ParseError::InvalidUTF8 {
            context: "Fetch::parse_payload(track_name)",
            details: e.to_string(),
          })?;
        let start_location = Location::deserialize(payload)?;
        let end_location = Location::deserialize(payload)?;

        standalone_fetch_props = Some(StandAloneFetchProps {
          track_namespace,
          track_name,
          start_location,
          end_location,
        });
      }
    }

    let param_count_u64 = payload.get_vi()?;
    let param_count: usize =
      param_count_u64
        .try_into()
        .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
          context: "Fetch::deserialize(param_count)",
          from_type: "u64",
          to_type: "usize",
          details: e.to_string(),
        })?;

    let mut parameters = Vec::with_capacity(param_count);
    for _ in 0..param_count {
      let param = KeyValuePair::deserialize(payload)?;
      parameters.push(param);
    }

    Ok(Box::new(Fetch {
      request_id,
      subscriber_priority,
      group_order,
      fetch_type,
      standalone_fetch_props,
      joining_fetch_props,
      parameters,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::Fetch
  }
}

#[cfg(test)]
mod tests {

  use super::*;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let request_id = 161803;
    let subscriber_priority = 15u8;
    let group_order = GroupOrder::Descending;
    let fetch_type = FetchType::AbsoluteFetch;
    let joining_fetch_props = JoiningFetchProps {
      joining_request_id: 119,
      joining_start: 73,
    };
    let parameters = vec![
      KeyValuePair::try_new_varint(4444, 12321).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"fetch me ok")).unwrap(),
    ];
    let fetch = Fetch {
      request_id,
      subscriber_priority,
      group_order,
      fetch_type,
      standalone_fetch_props: None,
      joining_fetch_props: Some(joining_fetch_props),
      parameters,
    };

    let mut buf = fetch.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Fetch as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = Fetch::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, fetch);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let request_id = 161803;
    let subscriber_priority = 15u8;
    let group_order = GroupOrder::Descending;
    let fetch_type = FetchType::AbsoluteFetch;
    let joining_fetch_props = JoiningFetchProps {
      joining_request_id: 119,
      joining_start: 73,
    };
    let parameters = vec![
      KeyValuePair::try_new_varint(4444, 12321).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"fetch me ok")).unwrap(),
    ];
    let fetch = Fetch {
      request_id,
      subscriber_priority,
      group_order,
      fetch_type,
      standalone_fetch_props: None,
      joining_fetch_props: Some(joining_fetch_props),
      parameters,
    };

    let serialized = fetch.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Fetch as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let deserialized = Fetch::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, fetch);
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let request_id = 161803;
    let subscriber_priority = 15u8;
    let group_order = GroupOrder::Descending;
    let fetch_type = FetchType::AbsoluteFetch;
    let joining_fetch_props = JoiningFetchProps {
      joining_request_id: 119,
      joining_start: 73,
    };
    let parameters = vec![
      KeyValuePair::try_new_varint(4444, 12321).unwrap(),
      KeyValuePair::try_new_bytes(1, Bytes::from_static(b"fetch me ok")).unwrap(),
    ];
    let fetch = Fetch {
      request_id,
      subscriber_priority,
      group_order,
      fetch_type,
      standalone_fetch_props: None,
      joining_fetch_props: Some(joining_fetch_props),
      parameters,
    };

    let mut buf = fetch.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::Fetch as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = Fetch::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
